# Praktikum Naïve Bayes

**Nama** : Christina Adelia Sitinjak  
**NIM** : 42324056

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-learn
iris = load_iris()

#### Penjelasan:
Kode ini untuk memanggil library yang dibutuhkan: `numpy` untuk hitungan array dan `pandas` untuk mengolah data dalam bentuk tabel. Dataset Iris tidak perlu diunduh dulu karena sudah tersedia di Scikit-Learn, tinggal dipanggil pakai `load_iris()`.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


#### Penjelasan:
Data Iris diubah jadi **DataFrame** supaya lebih gampang dibaca dan diolah. Kolom `Species` dibuat dari `iris.target_names[iris.target]`, mengubah angka 0/1/2 menjadi nama spesiesnya (setosa, versicolor, virginica).

Setelah itu data dipisah jadi `X` (4 fitur: sepal length, sepal width, petal length, petal width) sebagai input, dan `y` (kolom `Species`) sebagai target yang mau ditebak. Dari output terlihat ada 150 baris data bunga dengan 4 kolom fitur.

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

#### Penjelasan:
Kode ini untuk melihat isi `y`, yaitu label/jawaban dari tiap data. Hasilnya berupa nama spesies untuk 150 data, dengan 50 data pertama semuanya `setosa`, lalu berlanjut ke `versicolor`, dan diakhiri dengan `virginica`.

In [4]:
from sklearn.preprocessing import LabelEncoder

lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

#### Penjelasan:
Model di Scikit-Learn bekerja dengan angka, sedangkan label kita masih berupa teks. Jadi label diubah jadi angka pakai `LabelEncoder`, urut sesuai abjad: **setosa = 0**, **versicolor = 1**, **virginica = 2**.

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

#### Penjelasan:
Data dibagi jadi **80% data latih** (untuk model belajar menghitung rata-rata dan standar deviasi tiap fitur per kelas) dan **20% data uji** (untuk mengecek model dengan data yang belum pernah dilihat), jadi ada 120 data latih dan 30 data uji. `random_state=0` supaya pembagiannya selalu sama tiap kali dijalankan.

In [6]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model.fit(X_train, y_train)

#### Penjelasan:
Karena keempat fitur Iris berupa angka kontinu yang polanya mirip distribusi normal (lonceng), dipakai **GaussianNB**. Saat `fit()` dipanggil, model ini sebenarnya sedang menghitung **rata-rata** dan **standar deviasi** dari tiap fitur, untuk masing-masing kelas (setosa, versicolor, virginica). Nilai-nilai itu nanti dipakai untuk menghitung peluang saat memprediksi data baru.

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

#### Penjelasan:
`NB_model.predict(X_test)` dipakai untuk menebak kelas dari 30 data uji. Untuk tiap data, model menghitung peluang data itu masuk ke masing-masing kelas, lalu memilih kelas dengan peluang paling besar. Hasilnya masih berupa angka (0/1/2) sesuai hasil encode tadi.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


#### Penjelasan:
Hasil prediksi (`result`) dibandingkan dengan jawaban asli (`y_test`) pakai `accuracy_score`, dan akurasinya **96,67%**. Artinya dari 30 data uji, 29 ditebak benar dan cuma 1 yang salah. Ini menunjukkan Naïve Bayes juga cukup bagus untuk dataset Iris, mirip seperti hasil KNN di praktikum sebelumnya.

In [ ]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\sitin_veptvir\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


#### Penjelasan:
Model dicoba untuk menebak satu data baru: `[5.1, 3.5, 1.4, 0.2]`. Hasil `predict()` masih berupa angka (0/1/2), jadi diubah lagi jadi nama spesies pakai `lab.inverse_transform()`. Hasilnya **setosa**.

Tapi kalau dijalankan, kode ini memunculkan **peringatan (warning)**, bukan error:
```
UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
```
Peringatan ini muncul karena saat `fit()`, data yang dipakai (`X_train`) berbentuk DataFrame yang punya nama kolom. Tapi saat `predict()`, `input_data` cuma berbentuk list biasa yang tidak punya nama kolom. Jadi ada ketidakcocokan bentuk data antara saat melatih dan saat memprediksi. Program tetap jalan dan hasilnya tetap benar, cuma Scikit-Learn mengingatkan supaya bentuk datanya dibuat konsisten.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


#### Penjelasan:
Supaya peringatan tadi hilang, `input_data` diubah dulu jadi **DataFrame** dengan nama kolom yang sama persis seperti `iris.feature_names`, sama seperti bentuk data yang dipakai saat `fit()`. Setelah bentuknya konsisten, prediksinya tetap **setosa**, tapi kali ini tanpa peringatan sama sekali.

Ini jadi kebiasaan yang baik: setiap kali mau memprediksi data baru, sebaiknya bentuknya disamakan dengan data yang dipakai untuk melatih model.

## Tugas

**1. Ikuti semua tahapan dalam modul ini.**
Semua langkah di atas sudah saya kerjakan urut dari awal sampai akhir.

**2. Prediksi data baru masuk ke species apa: `[7.3, 3.8, 4.6, 2.3]`**

In [11]:
# Prediksi data baru untuk tugas
input_tugas = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)

hasil_tugas = NB_model.predict(input_tugas)
label_tugas = lab.inverse_transform(hasil_tugas)

# Lihat peluang untuk tiap kelas
proba = NB_model.predict_proba(input_tugas)

print('Hasil prediksi untuk input baru adalah:', label_tugas)
print('Peluang tiap kelas (setosa, versicolor, virginica):', proba)

Hasil prediksi untuk input baru adalah: ['virginica']
Peluang tiap kelas (setosa, versicolor, virginica): [[4.83048542e-194 3.68081469e-007 9.99999632e-001]]


#### Penjelasan:
Sama seperti sebelumnya, input diubah dulu jadi DataFrame dengan nama kolom yang sesuai supaya tidak muncul peringatan. Selain `predict()`, saya juga pakai `predict_proba()` untuk melihat berapa besar peluang data ini masuk ke tiap kelas, jadi kelihatan seberapa yakin modelnya.

Hasilnya: input `[7.3, 3.8, 4.6, 2.3]` diprediksi masuk ke spesies **virginica**, dengan peluang hampir 100%. Ini masuk akal karena nilai `petal length` (4.6) dan `petal width` (2.3) pada input ini besar, dan dari data latih kita tahu virginica memang punya petal yang paling besar dibanding setosa dan versicolor.